File: algorithms/0056.merge_intervals/merge_intervals.rs
impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));

        let mut merge_intervals = vec![];
        let mut i = 0;
        while i < intervals.len() {
            let (first, mut last) = (intervals[i][0], intervals[i][1]);

            while i + 1 < intervals.len() && intervals[i + 1][0] <= last  {
                last = last.max(intervals[i + 1][1]);
                i += 1;
            }
            merge_intervals.push(vec![first, last]);
            i += 1;
        }
        merge_intervals
    }
}

File: algorithms/0056.merge_intervals/merge_intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const auto &a, const auto &b) {
            if (a[0] == b[0]) {
                return a[1] < b[1];
            }
            return a[0] < b[0];
        });
        vector<vector<int>> merge_intervals;
        for (int i = 0; i < intervals.size(); ++i) {
            int first = intervals[i][0];
            int last = intervals[i][1];
            while (i + 1 < intervals.size() && last >= intervals[i + 1][0]) {
                last = max(intervals[i + 1][1], last);
                ++i;
            }
            merge_intervals.push_back({first, last});
        }
        return merge_intervals;
    }
};